subsys/bluetooth/host/settings.c

```diff
@@ -33,75 +33,129 @@
+
+static int bt_settings_encode_key_no_addr(char *path, size_t path_size, const char *key)
+{
+	int err;
+
+	err = snprintk(path, path_size, "bt/%s", key);
+	if (err < 0) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+#else /* !CONFIG_BT_SETTINGS_USE_PRINTK */
+int bt_settings_encode_key(char *path, size_t path_size, const char *subsys,
+			   const bt_addr_le_t *addr, const char *key)
 {
 	size_t len = 3;
 
-	/* Skip if path_size is less than 3; strlen("bt/") */
+	/* path_size is less than or equal 3; strlen("bt/") */
+	if (path_size <= len) {
+		return -EINVAL;
+	}
+
+	/* Key format:
+	 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
+	 */
+	strcpy(path, "bt/");
+
+	/* Concatenate subsys as much the free space permits */
+	strncpy(&path[len], subsys, (path_size - len));
+
+	/* Postfix '/' if there is free space */
+	len = strlen(path);
+	if (len < path_size) {
+		path[len] = '/';
+		len++;
+	}
+
+	/* Concatenate addr as much the free space permits */
+	for (int8_t i = 5; i >= 0 && len < path_size; i--) {
+		len += bin2hex(&addr->a.val[i], 1, &path[len],
+			       path_size - len);
+	}
+
+	/* Postfix addr type if there is free space */
 	if (len < path_size) {
-		/* Key format:
-		 *  "bt/<subsys>/<addr><type>/<key>", "/<key>" is optional
+		/* Type can be either BT_ADDR_LE_PUBLIC or
+		 * BT_ADDR_LE_RANDOM (value 0 or 1)
 		 */
-		strcpy(path, "bt/");
-		strncpy(&path[len], subsys, path_size - len);
-		len = strlen(path);
-		if (len < path_size) {
-			path[len] = '/';
-			len++;
-		}
+		path[len] = '0' + addr->type;
+		len++;
+	}
 
-		for (int8_t i = 5; i >= 0 && len < path_size; i--) {
-			len += bin2hex(&addr->a.val[i], 1, &path[len],
-				       path_size - len);
-		}
+	/* Postfix '/' and concatenate key as much the free space permits */
+	if ((key != NULL) && (len < path_size)) {
+		path[len] = '/';
+		len++;
+		strncpy(&path[len], key, path_size - len);
+		len += strlen(&path[len]);
+	}
 
-		if (len < path_size) {
-			/* Type can be either BT_ADDR_LE_PUBLIC or
-			 * BT_ADDR_LE_RANDOM (value 0 or 1)
-			 */
-			path[len] = '0' + addr->type;
-			len++;
-		}
+	/* Insufficient path_size, include null termination */
+	if (len >= path_size) {
+		return -EINVAL;
+	}
 
-		if (key && len < path_size) {
-			path[len] = '/';
-			len++;
-			strncpy(&path[len], key, path_size - len);
-			len += strlen(&path[len]);
-		}
+	LOG_DBG("Encoded path %s", path);
 
-		if (len >= path_size) {
-			/* Truncate string */
-			path[path_size - 1] = '\0';
-		}
-	} else if (path_size > 0) {
-		*path = '\0';
+	return 0;
+}
+
+static int bt_settings_encode_key_no_addr(char *path, size_t path_size, const char *key)
+{
+	size_t len = 3;
+
+	/* path_size is less than or equal 3; strlen("bt/") */
+	if (path_size <= len) {
+		return -EINVAL;
+	}
+
+	/* Key format:
+	 *  "bt/<key>"
+	 */
+	strcpy(path, "bt/");
+
+	/* Concatenate key as much the free space permits */
+	strncpy(&path[len], key, path_size - len);
+	len = strlen(path);
```

Comment:
I guess it's technically possible to use the return value of `strncpy` to determine if `len >= path_size` without calling `strlen` for a minor optimization (making it an O(1) check rather than a O(n)). Something like 
```c
end = strncpy(...)
if (end - path >= path_size) { ... }
```

It's OK as is (although I'd still prefer the length checks at the beginning of the function before attempting to copy anything)